### ISSUE TITLE
Version 0.0.34 gives drastic different ouput when paired with vicuna.

### ISSUE THREAD

**Issue by JoshuaeKaiser** [Atome-FE/llama-node#43]
The same code when user with 0.0.34 versus 0.0.26 gives very weird output effects when paired with the vicuna 7 B model

0.0.26 works perfectly, so happy to stay on it, but whatever 0.0.34 shipped has broken something.

See example screens bellow. 

Prompt: "Count from 1 to 10"

0.0.26:

<img width="722" alt="Screenshot 2023-05-04 at 22 22 39" src="https://user-images.githubusercontent.com/17850490/236333019-a0853c55-a04f-42cb-acf6-cdd514ac9a1b.png">

0.0.34: 

<img width="765" alt="Screenshot 2023-05-04 at 22 21 58" src="https://user-images.githubusercontent.com/17850490/236333070-dca7180a-1f5b-498d-975b-e91fe342958a.png">

As you can see the model goes haywire, and this is the same across many different prompts.

**Comment by tchereau** [comment]
have you tried with llama.cpp ?
if the result is the same @hlhr202 can't fix it, and you have to wait for an update from llama.cpp

**Comment by hlhr202** [comment]
@JoshuaeKaiser Hi, llama sampling is not idempotent if you setup topk topp and nonzero temperature.ie. your output will never be same for same parameters(except that you set temperature to 0 but not sure if the backend has implemeted for this). could you provide an example for both with temp=0.1 topk=1 topp=0.1? which may generate more stable output.
additionally for vicuna please upgrade to v1.1 and use the new instruct template as described [here](https://github.com/lm-sys/FastChat/blob/main/docs/vicuna_weights_version.md#example-prompt-weight-v11)

**Comment by JoshuaeKaiser** [comment]
@hlhr202 Thank you for your detailed reply. I have managed to track down the issue, I was using the master prompt template the original docks specified instead of what your new example states shown bellow.

const prompt = `A chat between a user and an assistant.
USER: ${template}
ASSISTANT:`;

One observation, the new example does not specify the "stopSequence" property, which I have come to find is crucial when specifiying that the model stop, so "stopSequence": "USER:" should probably be added to that example, otherwise the model continues forever until it hits maxium output.

Couple of questions if you will indulge me. What do these properties denote when passed as config. I have added coments regarding my assumptions on some of them.

	nCtx: 1024, //Maximum input context size of the prompt
	nParts: -1,
	seed: 0,
	f16Kv: false,
	logitsAll: false,
	vocabOnly: false,
	useMlock: false,
	embedding: true, //Enables embedding mode so that you can generate embeddings using the specified methods
	useMmap: true,
	
	nThreads: 4,
	nTokPredict: 2048, //Maximum number of tokens the model will output
	topK: 40,
	topP: 0.1,
	
	
One further question. For running multiple instances of the model for example sticking the model behind a websocket api, do you just recomend calling the LLAMA constructor on every request and loading the model each time like below? I ask because this obviously adds time to the inference output. Wondering what the best approach is here to increase the speed at which inference starts

const llama = new LLama(LLamaCpp);

llama.load({path: model, ...config})

**Comment by hlhr202** [comment]
> @hlhr202 Thank you for your detailed reply. I have managed to track down the issue, I was using the master prompt template the original docks specified instead of what your new example states shown bellow.
> 
> const prompt = `A chat between a user and an assistant. USER: ${template} ASSISTANT:`;
> 
> One observation, the new example does not specify the "stopSequence" property, which I have come to find is crucial when specifiying that the model stop, so "stopSequence": "USER:" should probably be added to that example, otherwise the model continues forever until it hits maxium output.
> 
> Couple of questions if you will indulge me. What do these properties denote when passed as config. I have added coments regarding my assumptions on some of them.
> 
> ```
> nCtx: 1024, //Maximum input context size of the prompt
> nParts: -1,
> seed: 0,
> f16Kv: false,
> logitsAll: false,
> vocabOnly: false,
> useMlock: false,
> embedding: true, //Enables embedding mode so that you can generate embeddings using the specified methods
> useMmap: true,
> 
> nThreads: 4,
> nTokPredict: 2048, //Maximum number of tokens the model will output
> topK: 40,
> topP: 0.1,
> ```
> 
> One further question. For running multiple instances of the model for example sticking the model behind a websocket api, do you just recomend calling the LLAMA constructor on every request and loading the model each time like below? I ask because this obviously adds time to the inference output. Wondering what the best approach is here to increase the speed at which inference starts
> 
> const llama = new LLama(LLamaCpp);
> 
> llama.load({path: model, ...config})

@JoshuaeKaiser 
Since I havent implemented parallel inference yet, if you want to serve llama-node for multi-inference concurrency purpose, you should create multiple llama instances(which consumes n x memory). but if you are just serve for one inference session, you can reuse the loaded model. which means you can just load model once, and do multiple inferences later.
